File: python/cds_harness/schema/smt.py
"""Pydantic v2 mirror of ``cds_kernel::schema::smt``.

See ``crates/kernel/src/schema/smt.rs`` for the canonical Rust definition.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict


class LabelledAssertion(BaseModel):
    """A retractable, named SMT-LIBv2 assertion (subject to ``check-sat-assuming``)."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    label: str
    formula: str
    enabled: bool
    provenance: str | None = None


class SmtConstraintMatrix(BaseModel):
    """The SMT-LIBv2 program presented to Z3 / cvc5."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    schema_version: str
    logic: str
    theories: list[str]
    preamble: str
    assumptions: list[LabelledAssertion]

File: frontend/src/lib/schemas/smt.ts
/**
 * `SmtConstraintMatrix` — TS mirror of `crates/kernel/src/schema/smt.rs`.
 *
 * `provenance` is `Option<String>` on the Rust side without a
 * `skip_serializing_if`, so on the wire it surfaces as `null` (not absent)
 * when the assertion is kernel-synthesized. The TS shape matches with
 * `string | null` rather than `string | undefined`.
 */

import { asArray, asBool, asNullable, asObject, asString } from './parse';

export interface LabelledAssertion {
	label: string;
	formula: string;
	enabled: boolean;
	provenance: string | null;
}

export interface SmtConstraintMatrix {
	schema_version: string;
	logic: string;
	theories: string[];
	preamble: string;
	assumptions: LabelledAssertion[];
}

function parseAssertion(value: unknown, path: string): LabelledAssertion {
	const obj = asObject(value, path);
	return {
		label: asString(obj.label, `${path}.label`),
		formula: asString(obj.formula, `${path}.formula`),
		enabled: asBool(obj.enabled, `${path}.enabled`),
		provenance: asNullable(obj.provenance, `${path}.provenance`, asString)
	};
}

export function parseSmtConstraintMatrix(value: unknown): SmtConstraintMatrix {
	const obj = asObject(value, 'SmtConstraintMatrix');
	return {
		schema_version: asString(obj.schema_version, 'SmtConstraintMatrix.schema_version'),
		logic: asString(obj.logic, 'SmtConstraintMatrix.logic'),
		theories: asArray(obj.theories, 'SmtConstraintMatrix.theories', asString),
		preamble: asString(obj.preamble, 'SmtConstraintMatrix.preamble'),
		assumptions: asArray(obj.assumptions, 'SmtConstraintMatrix.assumptions', parseAssertion)
	};
}
